### PR TITLE
OC Proto to Jaeger Thrift translation

### PR DESCRIPTION
**Pull request by pjanotti**
Add a translator from OC-Proto to Jaeger Thrift.

As usual there are some questions about the translation,
they should be marked with TODOs but we may want to
follow up with some discussion about the general rules in
translation (e.g.: preserve key-value info even if it maps to
some property).

### CODE REVIEW

File: translator/trace/jaegerthrift_to_protospan.go [hunk 1]
```diff
@@ -227,15 +228,12 @@ func jtagsToAttributes(tags []*jaeger.Tag) (string, tracepb.Span_SpanKind, *trac
 		case "http.status_code", "status.code": // It is expected to be an int
 			statusCodePtr = new(int32)
 			*statusCodePtr = int32(tag.GetVLong())
-			continue
```
Comment (odeke-em) [inline translator/trace/jaegerthrift_to_protospan.go]: Why did you remove this? If we are capturing the status code as per the values declared in Lines 210 and 211 we don't need to add these as attributes.

File: translator/trace/jaegerthrift_to_protospan.go [hunk 2]
```diff
@@ -227,15 +228,12 @@ func jtagsToAttributes(tags []*jaeger.Tag) (string, tracepb.Span_SpanKind, *trac
...
 
 		case "http.status_message", "status.message":
 			statusMessage = tag.GetVStr()
-			continue
```
Comment (odeke-em) [inline translator/trace/jaegerthrift_to_protospan.go]: Ditto about this continue being removed.

File: translator/trace/jaegerthrift_to_protospan_test.go [hunk 3]
```diff
@@ -208,6 +208,20 @@ func TestConservativeConversions(t *testing.T) {
 						Code:    5,
 						Message: "cache miss",
 					},
+					Attributes: &tracepb.Span_Attributes{
+						AttributeMap: map[string]*tracepb.AttributeValue{
```
Comment (odeke-em) [inline translator/trace/jaegerthrift_to_protospan_test.go]: We want to capture the status code and not capture it as an attribute otherwise there is no point in setting the status above. What's your rationale for this change?

File: translator/trace/jaegerthrift_to_protospan_test.go [hunk 4]
```diff
@@ -217,6 +231,20 @@ func TestConservativeConversions(t *testing.T) {
 						Code:    13,
 						Message: "proxy crashed",
 					},
+					Attributes: &tracepb.Span_Attributes{
+						AttributeMap: map[string]*tracepb.AttributeValue{
+							"status.code": {
```
Comment (odeke-em) [inline translator/trace/jaegerthrift_to_protospan_test.go]: I don't think we want these attributes as they are, they should have been translated in status.

File: translator/trace/protospan_to_jaegerthrift.go [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+	// Pre-allocate assuming that few, if any spans, are nil.
+	jSpans := make([]*jaeger.Span, 0, len(ocSpans))
+	for _, ocSpan := range ocSpans {
+		startTime := timestampToEpochMicroseconds(ocSpan.StartTime)
+		jSpan := &jaeger.Span{
+			SpanId:        ocIDBytesToJaegerID(ocSpan.SpanId),
+			ParentSpanId:  ocIDBytesToJaegerID(ocSpan.ParentSpanId),
+			OperationName: truncableStringToStr(ocSpan.Name),
+			References:    ocLinksToJaegerReferences(ocSpan.Links),
+			// Flags: TODO (@pjanotti) at first nothing matches to it.
```
Comment (odeke-em) [inline translator/trace/protospan_to_jaegerthrift.go]: I don't understand this comment.

File: translator/trace/protospan_to_jaegerthrift.go [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+func truncableStringToStr(ts *tracepb.TruncatableString) string {
+	if ts == nil {
+		return ""
+	}
+	return ts.Value
+}
+
+func traceIDBytesToLowAndHigh(traceID []byte) (traceIDLow, traceIDHigh int64) {
+	if len(traceID) != 16 {
+		return
```
Comment (odeke-em) [inline translator/trace/protospan_to_jaegerthrift.go]: It would be useful to return an error instead of silently passing as TraceID is a very important field.

File: translator/trace/protospan_to_jaegerthrift.go [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+	}
+	traceIDHigh = int64(binary.BigEndian.Uint64(traceID[:8]))
+	traceIDLow = int64(binary.BigEndian.Uint64(traceID[8:]))
+	return
+}
+
+func ocIDBytesToJaegerID(b []byte) (id int64) {
+	if len(b) != 8 {
+		return
```
Comment (odeke-em) [inline translator/trace/protospan_to_jaegerthrift.go]: Same here it'd be nice to return an error.

File: translator/trace/protospan_to_jaegerthrift_test.go [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+
+	commonpb "github.com/census-instrumentation/opencensus-proto/gen-go/agent/common/v1"
+	agenttracepb "github.com/census-instrumentation/opencensus-proto/gen-go/agent/trace/v1"
+	tracepb "github.com/census-instrumentation/opencensus-proto/gen-go/trace/v1"
+	"github.com/census-instrumentation/opencensus-service/internal/testutils"
+	"github.com/golang/protobuf/ptypes/timestamp"
+	"github.com/jaegertracing/jaeger/thrift-gen/jaeger"
+)
+
+func TestTraceIDRoundTripJaegerOCProtoJaeger(t *testing.T) {
```
Comment (odeke-em) [inline translator/trace/protospan_to_jaegerthrift_test.go]: Perhaps make the prefix `TestJaegerFromOCProto` otherwise the test's name becomes hard to parse since the reader has to examine the suffix instead of prefix which humans find easier scanning.

File: translator/trace/protospan_to_jaegerthrift_test.go [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+				"a.binary": "AQIDBAMCAQ==",
+				"a.bool":   "true",
+				"a.double": "1234.56789",
+				"a.long":   "123456789",
+				"ip":       "10.53.69.61",
+			},
+		},
+		Spans: []*tracepb.Span{
+			{
+				TraceId:      base64ToBytes("AAAAAAAAAABSlpqJVVcaPw=="),
```
Comment (odeke-em) [inline translator/trace/protospan_to_jaegerthrift_test.go]: Why are we using base64? I think just plain bytes are easier to parse and easy to examine as a human and for example this is 24 bytes which becomes impossible to eyeball and catch conversion bugs. Please take a look at other tests for how we use plain `[]byte{0x01,....}`.

File: translator/trace/protospan_to_jaegerthrift_test.go [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+		}
+
+		gjson, _ := json.Marshal(gotJBatch)
+		wjson, _ := json.Marshal(wantJBatch)
+		gjsonStr := testutils.GenerateNormalizedJSON(string(gjson))
+		wjsonStr := testutils.GenerateNormalizedJSON(string(wjson))
+		if gjsonStr != wjsonStr {
+			t.Errorf("OC Proto to Jaeger Thrift failed.\nGot:\n%s\nWant:\n%s\n", gjsonStr, wjsonStr)
+			t.Errorf("Saving formated jsons at ./testdata/ to help with further investigation")
+			testutils.SaveAsFormattedJSON(fmt.Sprintf("./testdata/failure_oc_to_jaeger_got_%02d.json", i), gotJBatch)
```
Comment (odeke-em) [inline translator/trace/protospan_to_jaegerthrift_test.go]: I don't think we want tests writing arbitrary JSON to file systems. Please remove this unless we clear these files on exit but this might not be pleasant to users who clone our project and get a bug.

File: translator/trace/jaegerthrift_to_protospan.go (hunk 1 above)
Comment (pjanotti) [reply translator/trace/jaegerthrift_to_protospan.go]: The issue is that the OC spec indicates that we should be actually mapping the values to the OC ones (see https://github.com/census-instrumentation/opencensus-specs/blob/master/trace/HTTP.md#mapping-from-http-status-codes-to-trace-status-codes), with that it is not possible to restore the original keys from OC -> Jaeger, so dropping the key-value pairs is losing information that some systems in the backend may be relying upon.

At this stage it seems that a better general approach is: it is ok/desirable to lift information from known key-value pairs but we should preserve original pairs since in the backend there may be systems relying on these tags "http.status_code" seems a very likely one for this case.

File: translator/trace/protospan_to_jaegerthrift.go (hunk 5 above)
Comment (pjanotti) [reply translator/trace/protospan_to_jaegerthrift.go]: Jaeger uses this as a bit flags but I didn't find the related/equivalent in OC proto, so it seems that the info that they carry is lost. I will reword it.

File: translator/trace/protospan_to_jaegerthrift_test.go (hunk 10 above)
Comment (pjanotti) [reply translator/trace/protospan_to_jaegerthrift_test.go]: If a change breaks the test this should speedup the fixing it. It will be visible as untracked file together with the other test files.

File: translator/trace/jaegerthrift_to_protospan.go (hunk 1 above)
Comment (odeke-em) [reply translator/trace/jaegerthrift_to_protospan.go]: Oh okay I see, thanks for the references. As a minor point, in that case we need to then separate `http.status_code` and translate them to OpenCensus status codes.

File: translator/trace/protospan_to_jaegerthrift_test.go (hunk 10 above)
Comment (odeke-em) [reply translator/trace/protospan_to_jaegerthrift_test.go]: I am uncomfortable with writing arbitrary files and leaving them on user's disks. In many projects I've encountered this as a discomfort and actually last week, I helped clean up such in the Go project from a stray write https://github.com/golang/go/issues/28781. We can print them to stdout.

File: translator/trace/protospan_to_jaegerthrift.go [hunk 11]
```diff
@@ -0,0 +1,360 @@
...
+		hostTag := &jaeger.Tag{
+			Key:   "hostname",
+			VType: jaeger.TagType_STRING,
+			VStr:  &node.Identifier.HostName,
+		}
+		jTags = append(jTags, hostTag)
+	}
+
+	// Add OpenCensus library information as tags if available
+	ocLib := node.LibraryInfo
```
Comment (odeke-em) [inline translator/trace/protospan_to_jaegerthrift.go]: Nice!

File: translator/trace/protospan_to_jaegerthrift.go [hunk 12]
```diff
@@ -0,0 +1,360 @@
...
+			jTag := &jaeger.Tag{
+				Key:   k,
+				VType: jaeger.TagType_STRING,
+				VStr:  &str,
+			}
+			jTags = append(jTags, jTag)
+		}
+	}
+
+	if node.Identifier != nil && node.Identifier.HostName != "" {
```
Comment (odeke-em) [inline translator/trace/protospan_to_jaegerthrift.go]: Cool! How about Pid and StartTimestamp?

Field|Reference
---|---
Pid|[ProcessIdentifier.Pid](https://godoc.org/github.com/census-instrumentation/opencensus-proto/gen-go/agent/common/v1#ProcessIdentifier.Pid)
StartTimestamp|[ProcessIdentifier.StartTimestamp](https://godoc.org/github.com/census-instrumentation/opencensus-proto/gen-go/agent/common/v1#ProcessIdentifier.StartTimestamp)

File: translator/trace/jaegerthrift_to_protospan_test.go [hunk 13]
```diff
@@ -208,6 +208,20 @@ func TestConservativeConversions(t *testing.T) {
 						Code:    5,
 						Message: "cache miss",
 					},
+					Attributes: &tracepb.Span_Attributes{
+						AttributeMap: map[string]*tracepb.AttributeValue{
+							"http.status_code": {
+								Value: &tracepb.AttributeValue_IntValue{
+									IntValue: 5,
```
Comment (odeke-em) [inline translator/trace/jaegerthrift_to_protospan_test.go]: In that case then let's use real status codes i.e. >=1XX

File: translator/trace/jaegerthrift_to_protospan.go (hunk 1 above)
Comment (pjanotti) [reply translator/trace/jaegerthrift_to_protospan.go]: >  in that case we need to then separate http.status_code and translate them to OpenCensus status codes.

Agreed, created #238 to track it.

File: translator/trace/protospan_to_jaegerthrift_test.go (hunk 10 above)
Comment (pjanotti) [reply translator/trace/protospan_to_jaegerthrift_test.go]: Think of these files as penalty for someone breaking the tests 😃 , but that is fine, I can quickly add this code if I want to compare the files. I will remove it and the related function.

File: translator/trace/protospan_to_jaegerthrift.go (hunk 12 above)
Comment (pjanotti) [reply translator/trace/protospan_to_jaegerthrift.go]: Added.

File: translator/trace/jaegerthrift_to_protospan_test.go (hunk 13 above)
Comment (pjanotti) [reply translator/trace/jaegerthrift_to_protospan_test.go]: Done.

File: translator/trace/protospan_to_jaegerthrift_test.go (hunk 10 above)
Comment (odeke-em) [reply translator/trace/protospan_to_jaegerthrift_test.go]: LOLOL

File: translator/trace/protospan_to_jaegerthrift.go [hunk 14]
```diff
@@ -0,0 +1,381 @@
...
+
+	return ocSpanAttributesToJaegerTags(annotation.Attributes)
+}
+
+func ocMessageEventToJaegerTags(msgEvent *tracepb.Span_TimeEvent_MessageEvent) []*jaeger.Tag {
+	if msgEvent == nil {
+		return nil
+	}
+
+	// TODO: (@pjanotti) Not clear how to map those to Jaeger, perhaps some OpenTracing tags...
```
Comment (odeke-em) [inline translator/trace/protospan_to_jaegerthrift.go]: I believe those map as OpenTracing "Span Logs", please perhaps file an issue otherwise we'll be losing some critical information :)

File: translator/trace/protospan_to_jaegerthrift.go [hunk 15]
```diff
@@ -0,0 +1,381 @@
...
+
+	return jLogs
+}
+
+func ocAnnotationToJagerTags(annotation *tracepb.Span_TimeEvent_Annotation) []*jaeger.Tag {
+	if annotation == nil {
+		return nil
+	}
+
+	// TODO: (@pjanotti) what about Description? Does it fit as another tag?
```
Comment (odeke-em) [inline translator/trace/protospan_to_jaegerthrift.go]: Please see my comment about MessageEventToJaegerTags being "Span Logs" for Jaeger

File: translator/trace/protospan_to_jaegerthrift.go [hunk 16]
```diff
@@ -0,0 +1,381 @@
...
+		}
+		jReferences, err := ocLinksToJaegerReferences(ocSpan.Links)
+		if err != nil {
+			return nil, fmt.Errorf("Error converting OC links to Jaeger references: %v", err)
+		}
+		spanID, err := ocIDBytesToJaegerID(ocSpan.SpanId)
+		if err != nil {
+			return nil, fmt.Errorf("OC span has invalid span ID: %v", err)
+		}
+		parentSpanID, err := ocIDBytesToJaegerID(ocSpan.ParentSpanId)
```
[the remark this answers is not in the document]
Comment (pjanotti) [reply translator/trace/protospan_to_jaegerthrift.go]: Since there will be special cases for the IDs on the function that understand the difference between ParentSpanId, SpanId and so on, I will do a bit different but overall with same intended effect.

File: translator/trace/protospan_to_jaegerthrift.go (hunk 14 above)
Comment (odeke-em) [reply translator/trace/protospan_to_jaegerthrift.go]: Perfect, thanks!
